Replace old Android client with new Air client

fortuna opened the pull request:
I'm sending the PR so that others can contribute to the new client.
We can also keep developing in a separate client if you prefer.

<!-- Reviewable:start -->

---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/uproxy/uproxy/2802)
<!-- Reviewable:end -->

trevj commented:
OK, cool - it's a start. Looking forward to building on this and removing some more of the hacks that were keeping the UI running. I had some comments on some parts of the code that aren't likely to be throwaway.

fortuna commented on src/cca/scripts/cloud_socks_proxy_server.ts:
```diff
@@ -0,0 +1,88 @@
+    }
+  }
+
+  public remoteIpAddress(): string {
```
Could you clarify?
That's a basic Typescript construct. I don't see why we would ban it.

trevj replied:
Sorry for the confusion. As discussed, while there are subtle differences between the following two code snippets, I'm struggling to think of good examples:
```ts
class Greeter {
    constructor(private message: string) {}
    getMessage() {
        return this.message;
    }
}
```
```ts
class Greeter {
    constructor(private message: string) {}
    getMessage = () => {
        return this.message;
    }
}
```
Additionally, the `only-arrow-functions` tslint function I'm working to incorporate **does not affect class methods**.

So, this is A-OK.

fortuna commented on src/cca/scripts/background.ts:
```diff
@@ -0,0 +1,41 @@
+var browserApi :CordovaBrowserApi = new CordovaBrowserApi();
+
+console.log('Loading core');
+export var uProxyAppChannel = freedom(
```
This is used by `CordovaCoreConnector`:

    <any>bgPage).ui_context.uProxyAppChannel

trevj replied:
Thanks! Worth adding a comment!

fortuna commented on src/cca/scripts/background.ts:
```diff
@@ -0,0 +1,41 @@
+export var moduleName = 'uProxy App Top Level';
+
+// TODO(fortuna): Move browser api logic directly here. No need for the extra layer.
+var browserApi :CordovaBrowserApi = new CordovaBrowserApi();
```
Done

fortuna commented on src/cca/scripts/background.ts:
```diff
@@ -0,0 +1,41 @@
+export interface OnEmitModuleFactory extends
+  freedom.FreedomModuleFactoryManager<OnEmitModule> {};
+
+export var moduleName = 'uProxy App Top Level';
```
It seems unused. I've deleted it now.

fortuna commented on src/cca/model/vpn_device.ts:
```diff
@@ -0,0 +1,9 @@
+export interface VpnDevice {
+  // Starts the VPN, redirecting the traffic to a Socks proxy running on the given port.
+  // You can pass a callback to be notified when it gets disconnected.
+  start(port: number, onDisconnect: (msg: string) => void) : Promise<string>;
```
We can't, because we always force localhost. I'd pass an `Endpoint` if we supported any IP.

fortuna commented on src/cca/model/socks_proxy_server.ts:
```diff
@@ -0,0 +1,18 @@
+  stop(): Promise<void>;
+}
+
+interface SocksProxyServerRepository {
```
Deleted. It was unused.

fortuna commented:
Hold on further review. I'm making some other fundamental changes that will move a bunch of things around.

fortuna commented:
All right @trevj  and @bemasc , this is ready for review again.

The main changes I made since last review:
* Made one button to connect, like in the final product (rather than one for the socks proxy and one for the vpn)
* Added the possibility to add multiple servers.
* Introduced the product concept of a "Server", which encapsulates the socks proxy and vpn connection details

Screenshot:
![image](https://cloud.githubusercontent.com/assets/113565/20150750/6f333b72-a685-11e6-8797-1ca3784f264d.png)

trevj commented:
Nice, nice, nice. I built and am playing around with it on my device. One thing so far: URL capture is not working for me. In the console I see:
```
[Background] Url is [null]
```
and
```
[App] Url: null
```

trevj commented on src/cca/scripts/background.ts:
```diff
@@ -0,0 +1,61 @@
+});
+
+// Register for url intents.
+Promise.all([serversListPagePromise, intents.GetGlobalIntentInterceptor()])
```
It's never nice to break just before a `.`. You can fit it within 80 characters if you do:
```ts
Promise.all([serversListPagePromise,
    intents.GetGlobalIntentInterceptor()]).then(([serverListPage,
    intentInterceptor]) => {
  intentInterceptor.addIntentListener((url: string) => {
    console.debug(`[App] Url: ${url}`);
    serverListPage.enterAccessCode(url);
  });
});
```

fortuna replied:
Done

trevj commented on src/cca/scripts/background.ts:
```diff
@@ -0,0 +1,61 @@
+});
+
+// Create UI.
+let serversListPagePromise: Promise<ServerListPage> = new Promise((resolve, reject) => {
```
There's a mix in this PR of `new Promise((F, R))` and `new Promise((resolve, reject))`. The style guide doesn't mention it but I think the first looks a little better and it's what most of the rest of our code does, too.

fortuna replied:
The Google style guides usually say you should avoid abbreviations ([example](https://google.github.io/styleguide/cppguide.html#General_Naming_Rules)). With the long names, the meaning is clear. In fact, when I first came to this code, it wasn't clear to me what `F` and `R` where. My first guess was Fail and Resolve, which was wrong and I had to look at the documentation to check what they actually were.

For that reason, I'd favor making the code clearer and more explicit over making it look good. I'm happy to change to something else that is clear. In other places of the code I've seen `fulfill` and `reject`. But I can do `F` and `R` if you feel strongly about it.
